### ISSUE TITLE
arrayfire using C++ for my existing project on visual studio 2012

### ISSUE THREAD

**Issue by kwasif** [arrayfire/arrayfire#2027]
I plan to integrate my current C code with arrayfire. I want to transfer my host array to device compute its absolute value and transfer it back. My code compiles but while running i get an error that var1 is not initialized. Any corrections? I wanted to ensure if i am using the right functions. Appreciate any help. Not much support for Arayfire C library,
float h[]={-1.5};
af_array var1;
af_array var2;
af_write_array (var1, h,sizeof(float), afHost);  // writing host to device 
af_abs(&var2,var1);
af_get_data_ptr(h,var2);

**Comment by 9prady9** [comment]
@kwasif You should look at the function `af_create_array`. `af_write_array` is used to copy given number of bytes from either a `af_array` or host pointer to another `af_array` whose allocation has already been done a.k.a it has been created previously using `af_create_array`. Hence you are getting error that `var1` is not initialized.

We strongly recommend using the C++ API for ArrayFire and largely use the C-API while providing the bindings to other programming languages like python, rust etc. Hence, all the tutorials and documentation examples use C++ API.

Having said that, if you still want to use the C API, please note that both C & C++ API are both documented(purpose of the function and the parameters) properly.

**Comment by 9prady9** [comment]
@kwasif Also, if you wanted to have the attention of @pavanky , please do so in a comment, not in the title of the issue.

**Comment by kwasif** [comment]
Thanks @9prady9 . I was able to integrate arrayfire kernel in my existing application which runs on visual studio 2012. But when i run the application in debug mode i dont get any display for info() or print_af statements. Eventually i am getting the right answer using af_kernel but it takes almost 20 sec compared to 0.08 sec with a CPU implementation.  I have tried using sync n tried to vectorize my computations. Appreciate any suggestions from your end.

**Comment by 9prady9** [comment]
On Windows, I have seen some times that DEBUG mode executions run pretty slow compared to Release mode. As far as the info() call output, it shouldn't matter if you are running in release or debug. Are you sure your application is console program according windows ?  

Which API are you using, C/C++ ? It isn't clear from your comment which API you are using. If you are printing an `af_array` you would have to call `af_print_array(af_array_variable_name_goes_here)`. If it is C++ API, then you have to pass an `af::array` object to `af_print(af::array_variable_name_here)`.

**Comment by kwasif** [comment]
I am using C++ API.  I already tried af_print() as well still no output. It is a mfc application but i am able to see other things in the output console. 
 
I ran it in release mode as well it takes around 11 sec but not even comparable to the CPU performance.
Given below is my algorithm. which is mainly taking the template around the non zero pixels positions in image to compute difference with with selective pixels of template. 
Input
Array image_af, edgeStrength_af, Img_binary with image width,height as I_w, I_h
Nz be number of non zero pixels in binary image.
Global_bin_af has histogram bin based on the diff score. 
Array current_template_x,current_template_y,current_template_val having x,y coordinate and its corresponding value. Size of this array is mapLength (76) for a template with width*height (14*14)
template_id=1
Function
for (thsIndex : all non zero pixels (x,y) img_binary)

if (x-w/2>-1 && x+w/2 <I_width && y-h/2>-1 && y+h/2 <I_height)

// vectorization of 76 template pixels
 array current_indices=(current_template_y+start_y+1)*width+current_template_y+start_y+1; 
array current_indices_val=lookup(image_af,current_indices,0); 
 array current_indices_edge=lookup(edgeStrength_af,current_indices,0); // float
array diff_af= (af::abs(current_indices_val-current_template_val))*0.1;
 diff_af=(current_indices_edge>0.1)*diff_af+(current_indices_edge<=0.1)*36;
diff_af= diff_af.as(u32);
array partial_score=lookup(Global_bin_af,diff_af,0);
scores_perpoint(thsIndex,template_id)=af::sum(partial_score)+0.75*mapLength;	

I feel i have vectorized my algorithm to suit array fire. Am i missing anything here?

**Comment by 9prady9** [comment]
How are you timing the ArrayFire code execution ? Have you gone through [this](http://arrayfire.org/docs/timing.htm) documentation ? First  call to any function incurrs additional device warm up time and the first iteration run time should be averaged out by running the ArrayFire code multiple times. These details are explained in the above linked documentation tutorial about timing ArrayFire code.
I see two calls to `af::lookup` in the above, random elmenet indexing on GPU memory is not a performant operation. May be that is the reason you are not seeing any improvement. Is there any pattern to how you are current_indices `af::array` data is organized ? or is it purely random ?

**Comment by kwasif** [comment]
I have used the timer::start() functionality to measure the time. Tried calling the function multiple times as well by looping over the same template again. 
Unfortunately there is no pattern as such, these indices are random based on non zero pixel locations of image and template.  

The indices are organized in a single array but the accessing of those indices would have to fetch random locations.  

I tried commenting the  3 lookup statements , abs() and the thresholding step the runtime went from 23.4 sec to 8.6 sec. I was under the impression arrayfire would automatically store the template and input image in L2 cache(reduces clock cycles) since it involves random access. so in such a case lookup should not take much time.  Anything else i can try here.

**Comment by 9prady9** [comment]
@kwasif Can you please share a standalone source file with ArrayFire code running your algorithm that runs out of the box. Would be lot easier to understand your code and understand the bottleneck if any exists.

**Comment by kwasif** [comment]
Strangely on visual studio 2013 i get output options in command prompt while using info() where as i dont get any output while using vs2012. The run time i get in release mode is 13 sec for first iteration and 12.3 sec for second iteration. The cpu code takes 0.09 sec. Attached is the algo af code with relevant txt files to load a few arrays. Look forward to your reply. 
[test.zip](https://github.com/arrayfire/arrayfire/files/1613622/test.zip)

**Comment by 9prady9** [comment]
@kwasif How long does your test program run ? I ran the code and it seems to be running for a while now (both on CPU and CUDA) ?

**Comment by kwasif** [comment]
The cpu implementation is a part of application which takes close to 0.1 sec for the same computation. 
If i use af_cpu option in array fire it takes 14 sec and using af_opencl with gpu device(NVIDIA 940MX) it takes around 12 sec. I tried commenting the entire for loop over non zero points and noticed it takes 12 sec again for no computation. I have installed opencl sdk etc as well. 
This is a small test size where we match the template along the non zero image locations.

**Comment by 9prady9** [comment]
@kwasif something else is going on your test setup then. For me the test program you gave took a lot more than 14 seconds using CPU backend also. In any case, please take a note of previous comment and try to vectorize your code.

**Comment by kwasif** [comment]
Sure, I was doing vectorization with in the Nz for loop to compute diffeernce between 76 (Tz) points and accumulate the sum. 

 I shall try to vectorize Nz and transfer all the arrays to device memory. The Nz vector gives rise to Nz*Tz computations when mapped against Tz. If i have a gfor loop for each of Nz is it a good approach or if i replicate Nz vector Tz times and then perform vector based operation.

**Comment by 9prady9** [comment]
Either way is fine I think, whichever is more readable and understandable
for you. GFOR basically expands into a vectorized operation.

**Comment by kwasif** [comment]
@9prady9 I took the approach of vectorizing the entire computation. Used the following approach by tiling the non zero points (thsLength) to map onto the template points (76). Aggregated the result for all the points then. This approach is still 4 times slower than CPU. Any more suggestions? 

                 array current_indices=array(thsLength,76,u32);
		  array start_y2=tile(start_y,1,76);
		  array start_x2=tile(start_x,1,76);
		  array current_template_y2=tile(current_template_y,1,thsLength);
		  array current_template_x2=tile(current_template_x,1,thsLength);
		  array current_indices_val=array(thsLength,76,f32);
		 
		  array current_template_val2=tile(current_template_val,1,thsLength);
		   current_indices=(start_y2+transpose(current_template_y2)+1)*width +start_x2+transpose(current_template_x2)+1;
		
                   current_indices_val=image_af(current_indices);
		   af::dim4 new_dims(thsLength, 76);
		   current_indices_val=moddims(current_indices_val,new_dims);
		  
		  array diff_af=af::abs(current_indices_val-transpose(current_template_val2))*0.1;
		  array current_indices_edge=edgeStrength_af(current_indices);
		  current_indices_edge=moddims(current_indices_edge,new_dims);

		  diff_af=(current_indices_edge>0.1)*diff_af+(current_indices_edge<=0.1)*36;

		  diff_af=diff_af.as(u32);
		  array bin_score=Global_bin_af(diff_af);
		  bin_score=moddims(bin_score,new_dims);
		  array bin_score_sum=sum(bin_score,1)+0.75*mapLength;
		  af::min(val,idx,bin_score_sum,-1);

**Comment by kwasif** [comment]
I did a bit of breakdown on the timing and noticed the below step consumed almost 0.12 sec(half of total AF kernel time 0.23)
 array diff_af=af::abs(current_indices_val-transpose(current_template_val2))*0.1;

**Comment by kwasif** [comment]
Resolved, apparently it was loading afcpu.lib due to which i wasnt able to see much speedup. Now i am getting a speedup of 6-7 times.

**Comment by 9prady9** [comment]
Glad, you resolved!

**Comment by 9prady9** [comment]
Leaving a note here for any future references for users who visit this issue.
@kwasif You can also link to `af.lib` which is the unified interface to our library. The advantage of unified interface is that you can use you can set the required backend `setBackend()` call and don't need to worry about changing the linking library and recompile your code every time. You can also avoid problems as above where you link wrong backend.

**Comment by kwasif** [comment]
Sure noted. Just one last thing in my new integrated project which is now working fine i am unable to get any display output while using info() or af_print_mem_info() tried on vs12 n vs13 still no luck. Any particular reason?

**Comment by 9prady9** [comment]
Not sure why that would happen. I have used vs12 for another project a month ago and everything was printing fine.

Where are you trying to see the info() print statements ? 

Is your program a win64 console application or a Windows application ? In the later case, I am not sure how printf statements work out - as in where does their output send to. Could it be that ?

**Comment by kwasif** [comment]
Its a windows application but normal printf statements work fine. Before launching the kernel while setting backend parameters i was using info(). I plan to use the af_print_mem_info("mem",0); to checkout the memory consumption in each iteration. The code seems to crash when i go into next iteration of template with a different size from 76 to 100. Gives me this error af::exception at memory location 0x000000AE6AEF8620. I tried using eval() but still get the same exception.

**Comment by 9prady9** [comment]
I am not sure why info() is not printing anything. I can assure that af::info() is internally just bunch of printf statements, so I don't think it is an issue from ArrayFire code. Can you reproduce this print issue on any standalone example that is shipped with ArrayFire ?

As far as the memory related problem, Can you check if it is memory limitation problem ? That is you are running out of memory when changing the template size or something like that. On windows, you can launch task manager and check GPU memory usage on Windows 10.

**Comment by kwasif** [comment]
info() works well with Standalone array fire application. The problem comes when i try to insert an AF kernel into my project. 
The memory utilization on GPU goes to 2% only. So its not a case of short of memory. When i make the AF kernel call twice with different size it works fine but when i put a for loop with varying template size it gives me this error. ArrayFire Exception (Invalid input size:203). 

Size mismatch between input and output for the statement 
array current_indices=(start_y2+transpose(current_template_y2)+1)*width +start_x2+transpose(current_template_x2)+1;

Where as all the dimensions are perfect. It works in one loop but fails in another. Is there some conflict where it is not able to modify the size of a variable in subsequent for loop and removes them from AF memory only after completing of for loop.  I used eval() and free() at end of for loop but still get the same error

**Comment by 9prady9** [comment]
@kwasif I am afraid it is hard to say anything without code. Since the original question about which this issue has been answered, please close this and kindly open another issue with the code sample that can help us reproduce the memory issue you are facing. Unless, the problem can be reproduced in a standalone example, it is hard debug the code remotely with error messages alone.
